vispy/scene/widgets/grid.py

```diff
@@ -183,63 +288,238 @@ def __repr__(self):
+    def _add_gridding_width_constraints(solver, width_grid):
+        # access widths of one "y", different x
+        for ws in width_grid.T:
+            for w in ws[1:]:
+                solver.add_constraint(ws[0] == w, strength=REQUIRED)
+
+    @staticmethod
+    def _add_gridding_height_constraints(solver, height_grid):
+        # access heights of one "y"
+        for hs in height_grid.T:
+            for h in hs[1:]:
+                solver.add_constraint(hs[0] == h, strength=REQUIRED)
+
+    @staticmethod
+    def _add_stretch_constraints(solver, width_grid, height_grid,
+                                 grid_widgets, widget_grid):
+        xmax = len(height_grid)
+        ymax = len(width_grid)
+
+        stretch_widths = [[] for _ in range(0, ymax)]
+        stretch_heights = [[] for _ in range(0, xmax)]
+
+        for (_, val) in grid_widgets.items():
+            (y, x, ys, xs, widget) = val
+
+            for ws in width_grid[y:y+ys]:
+                total_w = np.sum(ws[x:x+xs])
+
+                for sw in stretch_widths[y:y+ys]:
+                    sw.append((total_w, widget.stretch[0]))
+
+            for hs in height_grid[x:x+xs]:
+                total_h = np.sum(hs[y:y+ys])
+
+                for sh in stretch_heights[x:x+xs]:
+                    sh.append((total_h, widget.stretch[1]))
+
+        for (x, xs) in enumerate(widget_grid):
+            for(y, widget) in enumerate(xs):
+                    if widget is None:
+                        stretch_widths[y].append((width_grid[y][x], 1))
+                        stretch_heights[x].append((height_grid[x][y], 1))
+
+        for sws in stretch_widths:
+            if len(sws) <= 1:
+                continue
+
+            comparator = sws[0][0] / sws[0][1]
+
+            for (stretch_term, stretch_val) in sws[1:]:
+                solver.add_constraint(comparator == stretch_term / stretch_val,
+                                      strength=WEAK)
+
+        for sws in stretch_heights:
+            if len(sws) <= 1:
+                continue
+
+            comparator = sws[0][0] / sws[0][1]
+
+            for (stretch_term, stretch_val) in sws[1:]:
+                solver.add_constraint(comparator == stretch_term / stretch_val,
+                                      strength=WEAK)
+
+    @staticmethod
+    def _add_widget_dim_constraints(solver, width_grid, height_grid,
+                                    var_w, var_h, grid_widgets):
+        assert(var_w is not None)
+        assert(var_h is not None)
+
+        for ws in width_grid:
+            for w in ws:
+                solver.add_constraint(w >= 0, strength=REQUIRED)
+
+        for hs in height_grid:
+            for h in hs:
+                solver.add_constraint(h >= 0, strength=REQUIRED)
+
+        for (_, val) in grid_widgets.items():
+            (y, x, ys, xs, widget) = val
+
+            for ws in width_grid[y:y+ys]:
+                total_w = np.sum(ws[x:x+xs])
+                # assert(total_w is not None)
```

Comment:
this is completely unexplainable by me - *adding* this assert somehow makes `examples/basics/plotting/plot.py` work, *without* tripping the assert itself.

However, without the assert,`plot.py` fails with a `NoneType` error. 

I don't understand how the `assert` could doesn't even throw an `AssertionError` or something. It simply works for no discernible reason. The only thing I can think of is some sort of race condition, but I can't see anything obvious like that in this case.

AFAIK, `np.sum` returns `0` if the array is empty so that shouldn't be a problem. [Docs here](http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.sum.html)